doc/developer-guide/internal-libraries/RPC.en.rst

```diff
@@ -0,0 +1,403 @@
+
+   .. member:: int data_len
+
+      Length in bytes of the message. 
+
+A RPC message is sent as a ``MgmtMessageHdr`` followed by the serialized data in bytes. 
+
+Read:
+=====
+
+1. Read the MgmtMessageHdr:
+
+   a. ``MgmtUtils.cc::mgmt_read_pipe(int fd, char *buf, int bytes_to_read)`` calls,
+
+   #. ``ink_sock.cc::read_socket(int s, char *buffer, int length)`` casts char* buffer to void*, calls, 
+
+   #. ``read(int fd, void *buf, size_t count)`` <- a C system call.
+
+#. Read the content:
+
+   a. data length is given in the header, 
+
+   #. read data_len bytes with the same proceedure as the MgmtMessageHdr.
+
+
+Write:
+======
+
+1. create a MgmtMessageHdr and populate it with the message information (msg id, raw data, raw data length) then, 
+
+#. append the MgmtMessageHdr with the raw data via memcpy. When it is to be written, 
+
+#. ``MgmtUtils.cc::mgmt_write_pipe(int fd, char *buf, int bytes_to_write)`` calls,
+
+#. ``ink_sock.cc::write_socket(int s, const char *buffer, int length)`` casts buffer to const void* and calls,
+
+#. ``write(int fildes, const void *buf, size_t nbytes)`` <- a C system call.
+
+
+==========================
+RPC API for remote clients
+==========================
+
+:ts:git:`NetworkMessage.cc` provides a set of APIs for remote clients to communicate with |TManager|. Generally, the workflow is that a remote client will send a request to |TManager| to preform some operation on |TServer|. 
+
+|TManager| will then send a response with the return value. The return value only indicates if the request was successfully propogated to |TServer|. Thus, there is no way currently for |TServer| to send information back to remote clients.
+
+.. function:: TSMgmtError send_mgmt_request(const mgmt_message_sender &snd, OpType optype, ...)
+.. function:: TSMgmtError send_mgmt_request(int fd, OpType optype, ...)
+
+   Send a request from a remote client to |TManager|.
+
+.. function:: TSMgmtError send_mgmt_response(int fd, OpType optype, ...)
+
+   Send a response from |TManager| to remote client.
+
+.. function:: TSMgmtError send_mgmt_error(int fd, OpType optype, TSMgmtError error)
+
+   Send err from |TManager| to remote client.
+
+.. function:: TSMgmtError recv_mgmt_request(void *buf, size_t buflen, OpType optype, ...)
+
+   Read request from remote client. Used by |TManager|.
+
+.. function:: TSMgmtError recv_mgmt_response(void *buf, size_t buflen, OpType optype, ...)
+  
+   Read response from |TManager|. Used by remote clients. 
+
+How |TCtl| sends messages to |TManager|
+=======================================
+
+This details how a remote client may use the ``NetworkMessage.cc`` interface to send messages to |TManager|
+
+.. class:: mgmt_message_sender 
+
+.. class:: mgmtapi_sender : public mgmt_message_sender
+
+   This class implements a mgmt_message_sender and specifies a function to send messages to |TManager|.
+
+.. function:: TSMgmtError mgmtapi_sender::send(void *msg, size_t msglen) const override;
```

Comment:
should I also indent these functions?

Reply:
Indenting is a scoping indicator. Functions in classes are indented to show they are part of the class. Global functions in general will not be indented. In this case, the `function` should be indented and inside a `class:: mgmtapi_sender` (unless that's a namespace).